src/main/java/pokemanager/App.java

```diff
@@ -1,71 +1,96 @@
+    private boolean run;
 
-    public App(InputStream in, PrintStream out, Box box, String storageFilename) {
+    public App(InputStream in,
+               PrintStream pw,
+               Box box,
+               String storageFilename) {
         this.reader = new BufferedReader(new InputStreamReader(in));
-        this.out = out;
+        this.pw = pw;
         this.box = box;
         this.storageFilename = storageFilename;
+        this.commands = buildCommands();
+        this.run = false;
     }
 
-    public void acceptInput() {
+    private static List<Command> buildCommands() {
+        return Arrays.asList(
+                new RetrieveCommand(),
+                new StoreCommand(),
+                new SaveCommand(),
+                new ExitCommand());
+    }
+
+    private Command findCommand(String command) {
+        for (Command c : commands) {
+            if (c.respondsTo(command)) {
+                return c;
+            }
+        }
+        return null;
+    }
+
+	public String getStoragePath() {
+		return storageFilename;
+	}
+
+	public Box getBox() {
+		return box;	
+	}
+	
+	public void acceptInput() {
         try {
             String line = reader.readLine();
             handleInputLine(line);
         } catch (Exception IOException) {}
     }
 
     private void outputStartMessage() {
-        out.println("Commands:\n'box' to see stored Pokemon" +
+        pw.println("Commands:\n'box' to see stored Pokemon" +
                 "\n'store SPECIES NICKNAME' to store a Pokemon" +
                 "\n'save' to save your stored Pokemon for next time\n");
     }
 
-    private void run() {
+    public void run() {
+        run = true;
         outputStartMessage();
-        while (true) {
+        while (run) {
             acceptInput();
         }
     }
 
+    public void exit() {
+        run = false;
+    }
+
     private void handleInputLine(String line) throws IOException {
-        if (line.startsWith("store ")) {
-            box.store(line.substring(6));
-            out.println("Stored!\n");
-        } else if (line.equals("save")) {
-            save();
-            out.println("Saved!\n");
-        } else if (line.equals("box")) {
-            out.println(getBoxContents());
+        Command command = findCommand(line);
+        if (command == null) {
+            pw.println("Please enter a valid command.\n");
         } else {
-            out.println("Please enter a valid command.\n");
+            command.execute(line, this);
         }
     }
 
-    private String getBoxContents() {
-        return box.retrieve();
-    }
-
-    private void save() throws IOException {
-        String contents = box.getDataString();
-        FileWriter fw = new FileWriter(storageFilename);
-        fw.write(contents);
-        fw.close();
-    }
-
     public static void main(String[] args) throws IOException {
         String storageFilename = "/Users/sam/Documents/pokemanager/data/data";
```

Comment:
The path shouldn't be hardcoded. If someone else tries to run it on their machine, it's going to be a problem!